From 0990264c85f3b0063fd75261bc1d37e7c81975da Mon Sep 17 00:00:00 2001
From: Kevin O'Connor <kevin@koconnor.net>
Date: Mon, 28 Sep 2026 21:30:21 -0400
Subject: [PATCH 2/6] timer_irq: Add new timer_dispatch_many_polling()

Add a new variant of timer_dispatch_many() for chips that do not
support hardware irqs, or do not support hardware irqs being enabled
during timer dispatch.  This new variant allows the boards to poll for
events if many timers are run back to back.

Signed-off-by: Kevin O'Connor <kevin@koconnor.net>
(cherry picked from commit 5233845b23bb1ed61dfebcd0fa972b5010e1db57)
---
 src/generic/timer_irq.c | 35 +++++++++++++++++++++++++++++++++++
 src/generic/timer_irq.h |  4 ++++
 2 files changed, 39 insertions(+)

diff --git a/src/generic/timer_irq.c b/src/generic/timer_irq.c
index 9f6b59aa3..96cf47375 100644
--- a/src/generic/timer_irq.c
+++ b/src/generic/timer_irq.c
@@ -69,6 +69,41 @@ timer_dispatch_many(void)
     }
 }
 
+// Same as above, but for boards without irqs (or where irqs aren't reentrant)
+uint32_t
+timer_dispatch_many_polling(void)
+{
+    uint32_t tru = timer_repeat_until;
+    for (;;) {
+        // Run the next software timer
+        uint32_t next = sched_timer_dispatch();
+
+        uint32_t now = timer_read_time();
+        int32_t diff = next - now;
+        if (diff > (int32_t)TIMER_MIN_TRY_TICKS)
+            // Schedule next timer normally.
+            return next;
+
+        if (unlikely(timer_is_before(tru, now))) {
+            // Check if there are too many repeat timers
+            if (diff < (int32_t)(-timer_from_us(1000)))
+                try_shutdown("Rescheduled timer in the past");
+            timer_dispatch_task_poll();
+            if (sched_check_set_tasks_busy()) {
+                timer_repeat_until = now + TIMER_REPEAT_TICKS;
+                return now + TIMER_DEFER_REPEAT_TICKS;
+            }
+            timer_repeat_until = tru = now + TIMER_REPEAT_TICKS;
+        }
+
+        // Next timer in the past or near future - wait for it to be ready
+        while (unlikely(diff > 0)) {
+            timer_dispatch_irq_poll();
+            diff = next - timer_read_time();
+        }
+    }
+}
+
 // Make sure timer_repeat_until doesn't wrap 32bit comparisons
 void
 timer_task(void)
diff --git a/src/generic/timer_irq.h b/src/generic/timer_irq.h
index 031f0d6b1..03d353cad 100644
--- a/src/generic/timer_irq.h
+++ b/src/generic/timer_irq.h
@@ -2,5 +2,9 @@
 #define __GENERIC_TIMER_IRQ_H
 
 uint32_t timer_dispatch_many(void);
+uint32_t timer_dispatch_many_polling(void);
+
+void timer_dispatch_irq_poll(void);
+void timer_dispatch_task_poll(void);
 
 #endif // timer_irq.h
-- 
2.47.3

